cmd/nvidia-device-plugin/prestart-hook.go

```diff
@@ -0,0 +1,283 @@
+	logCtx := getLogContextFromDevices(devices)
+
+	// check if any of the devices are already unhealthy
+	for _, device := range devices {
+		if device.Health != "Healthy" {
+			return fmt.Errorf("Device %s has health status %s", deviceString(device), device.Health)
+		}
+	}
+
+	// lock devices
+	for _, device := range devices {
+
+		preStartInfo, err := hook.getDevicePreStartInfo(ctx, device.ID, logCtx)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("PreStartHook(%s): locking device %s...\n", logCtx, deviceString(device))
+		if err := preStartInfo.lockDevice(ctx); err != nil {
+			return err
+		}
+		log.Printf("PreStartHook(%s): locked device %s\n", logCtx, deviceString(device))
+		defer preStartInfo.unlockDevice(ctx)
+	}
+
+	// run pre-start container command
+	unhealthy, err := hook.runPreStartContainerCommand(ctx, devices)
+	log.Printf("PreStartHook(%s): pre start container command completed. unhealthy:%v command-err:%v context-err:%v\n", logCtx, unhealthy, err, ctx.Err())
+
+	// if ctx deadline exceeded or command exited with non-zero exit code then add failure to all devices
+	// when we fail 5 times then mark device as unhealthy
+	if err == nil {
+		err = ctx.Err()
+	}
+
+	if err != nil {
+		log.Printf("PreStartHook(%s): Error running pre-start command due to error: %v\n", logCtx, err)
+		for _, device := range devices {
+			preStartInfo, err := hook.getDevicePreStartInfo(ctx, device.ID, logCtx)
+			if err != nil {
+				log.Printf("PreStartHook(%s): %v", logCtx, err)
+				continue
+			}
+
+			failures := preStartInfo.incPreStartFailures()
+			log.Printf("PreStartHook(%s): Failure count for device %s: %d\n", logCtx, device.ID, failures)
+			if failures > 4 {
+				markUnhealthy(device, unhealthyChan, logCtx)
+			}
+		}
+		return err
+	}
+
+	// if no timeout and exit code 0 then reset all pre-start failure counts
+	for _, device := range devices {
+		if preStartInfo, err := hook.getDevicePreStartInfo(ctx, device.ID, logCtx); err == nil {
+			preStartInfo.resetPreStartFailures()
+		} else {
+			log.Printf("PreStartHook(%s): Unable to failure reset count. Device not found: %v\n", logCtx, err)
+		}
+
+	}
+
+	// mark reported devices as unhealthy
+	for _, uuid := range unhealthy {
+		if device, err := findDevice(uuid, devices); err == nil {
+			markUnhealthy(device, unhealthyChan, logCtx)
+		} else {
+			log.Printf("PreStartHook(%s): Unable to mark device as unhealty. Device not found: %v\n", logCtx, err)
+		}
+	}
+
+	if len(unhealthy) > 0 {
+		return fmt.Errorf("Found unhealthy devices %s", strings.Join(unhealthy, " "))
+	}
+	return nil
+}
+
+func (hook *PreStartHook) runPreStartContainerCommand(ctx context.Context, devices []*rm.Device) (unhealthy []string, err error) {
+	if len(devices) == 0 {
+		return
+	}
+	logCtx := getLogContextFromDevices(devices)
+
+	cmd := hook.preStartCmd[0]
+	args := hook.preStartCmd[1:]
+	for _, device := range devices {
+		uuid := strings.TrimPrefix(device.ID, "GPU-")
```

Comment:
Does this mean that we don't support MIG devices?

This also adds a specific API for the prestart commands which is not documented.